README.md
# composure

![yodawg](https://i.imgflip.com/sig5d.jpg)

do this:

![do this](https://cloud.githubusercontent.com/assets/2379901/10447049/9f4aeab6-7135-11e5-90a2-776997f3fb8b.png)


get this:

![composure-grid](https://cloud.githubusercontent.com/assets/2379901/10570762/4ac2a6bc-75ec-11e5-960f-74199112cdf8.png)

Composure makes it easy to add and remove grid functionality as you need it, with lightweight atomic output that won't slow down your site. Composure gives you a css system where each class does one little thing - and that's it. Based on the ideas of atomic-css, but without the incomprehensible syntax. No cascading bullshit, and still flexible enough to generate most kinds of css frameworks with a bit a fiddling around.

By passing in a config object, you generate a different grid and gutter system. Great for use in production (the WIRED preset is more or less what WIRED uses) or even just for rapid prototyping.

Another perk - since the actual css is dynamically generated maintenance is easy. Need to change some global variables, or media queries? Just update the properties in the config object and you're good to go, no weird bower or npm workarounds or manual overrides needed.

All options set here via the global config get set as global stylus vavriables. I also throw in a few mixins/extends. Feel free to override / use them as needed, although you shouldn't really need to do this (just change the config!).

### Presets
Composure comes with a few default grid presets. All of them are mobile first, and support flexbox by default (with helper classes for re-ordering, row direction, etc). 

#### simple
is the default grid. If you don't pass in a config object or specify a different preset, simple's defaults will be used. `simple` is a 12 column, mobile first grid, with 3 breakpoints (mobile, tablet, desktop). There are no pushes or pulls, but there are omega classes. It also includes a basic margin and padding gutter system. It's a whopping 7kb, and I've some variation of this scheme in production for a long time. It's awesome.

#### WIRED
is based on the grid I put together at WIRED. It's a slightly more complicated version of simple. It has 18 columsn instead of 12. It has 4 breakpoints  (mobile, tablet, desktop, and wide). All breakpoints except wide have maxes instead of just mins, and it has a slightly more extensive gutter system that is the exact one used by WIRED (8, 16, 32, 64).

#### bootstrap
is based on bootstrap 4 by twitter. It's basically a copy of their grid system - only you can override any aspect of it you don't like! It has 12 columns, 5 breakpoints, uses rems for media queries but not padding, and uses their class names and structure how you would expect. Also has pushes/pulls. No gutter system by default.

#### foundation
is based on foundation 5 by zurb. It's basically a copy of their grid system - only you can override any aspect of it you don't like! It has 12 columns, 5 breakpoints, uses rems for everything, and uses their class names and structure how you would expect. Also has pushes/pulls. No gutter system by default.

More presets will come as I get to it, or if anyone wants to write one :)


### Options
Eventually I'd like to go through and create demos for all these options, but you can play around with them yourself:


```javascript
'use strict'

var gulp = require( 'gulp' )
var stylus = require( 'gulp-stylus' )
var mq = require( 'css-mqpacker' )
var postcss = require( 'gulp-postcss' )
var composure = require( 'composure-grid' )

gulp.task( 'stylus', function() {
  return gulp.src( './styl/main.styl' )
    .pipe( stylus( {
      compress: false,
      use: composure( {
      // no margin left, used with a left first spacer grid
        alpha: false,
      // how many columns to generate
        columns: 12,
      // base column class
        columnBaseClass: 'col',
      // should columns have padding?
        columnPadding: false,
      // prefix the column classes (col-tab-1 instead of tab-1)
        columnPrefix: '',
      // if outputting gutters, do you want to output each direction as well?
        directions: {
          t: 'top',
          r: 'right',
          b: 'bottom',
          l: 'left',
        },
      // hide/show classes for each defined breakpoint
        displayClasses: true,
      // remove/reveal classes for each defined breakpoint
        displayOverrides: false,
      // by default use a flexbox based grid
        flexSupport: true,
      // include re-ordering classes for each breakpoint (order-1, tab-order-2, etc)
        flexOrder: true,
      // mobile first grid,
      // mark property that you will use as your mobile breakpoint
        gridBreakpoints: {
          mob: {
            max: 767,
            isMobile: true,
          },
          tab: {
            min: 768,
          },
          desk: {
            min: 1200,
          },
        },
      // margin and padding helper classes
      // pass in true for default, false for no gutters
      // values is used to hard code the gutter name and values
      // base/range/scale is used to generate a range of gutters
      // the example values object is what
      // the default base/range/scale combo creates
        gutters: {
          base: 10,
          range: 3,
          scale: 2,
          values: {
            0: 10,
            1: 20,
            2: 40,
          },
        },
      // include mobile column classes, default to just one grid on mobile
        includeMobile: false,
      // works with gridBreakpoints to designate the unit type of our min/max values
        mediaUnit: 'px',
      // include omega classes for zeroing out column spacers
        omega: true,
      // the default preset, you don't need to pass this in
      // when you change a preset the defaults just change
      // you can still override them
      // presets include WIRED style, bootstrap, foundation, with more coming maybe
        preset: 'simple',
      // push classes for columns (tab-push-3 etc)
        push: false,
      // pull classes for columns (tab-pull-3 etc)
        pull: false,
      // how wide the container class is
        siteWidth: 80,
      // should columns have default margin/padding? can be 'margin'|'padding'|false
        spacer: 'margin',
      // the amount of spacing to do
        spacerAmount: 20,
      // include spacing on the bottom as well
        spacerBottom: true,
      // what side of your columns do you want spacing:
      // 'left' || 'right' || 'both'
        spacerDirection: 'right',
      // the unit type of the spacer, can be whatever ('px'|'vw'|'rem'|'%' etc)
        spacerType: 'px',
      // specifies unit type for siteWidth property
        widthType: '%',
      } ),
    } ) )
    .pipe( postcss( [mq] ) )
    .pipe( gulp.dest( './css' ) )
} )
```



CHANGELOG.md
### 1.1.1

just updating docs and package.json info

### 1.1.0

add options for push and pull classes

add option for spacer direction (left, right, or both)

add foundation and bootstrap presets

### 1.0.1 -> 1.0.3

fix dumb bugs caused by lack of testing

### 1.0.0

release, support for simple and WIRED style grid systems


lib/presets/WIRED.js
'use strict'

/**
 * Module dependencies.
 */
var stylus = require( 'stylus' )
var _ = require( 'lodash' )
var path = require( 'path' )
var nodes = stylus.nodes

var wiredStyle = function( style, opts ) {
  // by default we support flexbox
  var flexBox = true
  if ( opts.flexSupport === false ) {
    flexBox = false
  }

  // by default we don't include column padding
  var padding = false
  if ( opts.columnPadding === true ) {
    padding =  new nodes.Literal( opts.columnPadding )
  }

  var defaultBreakpoints = {
    mob: {
      max: 767,
      isMobile: true,
    },
    tab: {
      min: 768,
      max: 960,
    },
    desk: {
      min: 961,
      max: 1199,
    },
    wide: {
      min: 1200,
    }
  }

  // these are the breakpoints we use to define the grid
  style.define( '$grid-breakpoints', opts.gridBreakpoints || defaultBreakpoints, true )
  style.define( '$media-unit', opts.mediaUnit || 'px' )

  var mob = [{}]
  if ( opts.gridBreakpoints ) {
    mob = _.filter( opts.gridBreakpoints, function(b) {
      return b.isMobile === true
    } )
  }
  else {
    mob = _.filter( defaultBreakpoints, function(b) {
      return b.isMobile === true
    } )
  }

  // we need to create a $mob-plus variable
  var unit = opts.mediaUnit || 'px'
  var min = mob[0].max + 1
  var mobPlus = '( min-width: ' + min + unit + ' )'
  style.define( '$mob-plus', mobPlus )

  // - GRID SETTINGS
  // how many columns to generate
  style.define( '$columns', opts.columns || 18 )
  // base column class
  style.define( '$column-base', opts.columnBaseClass || 'col' )

  // basic flex box support?
  style.define( '$flexbox-support', flexBox )
  // generate flexbox re-ordering classes
  style.define( '$flexbox-order', opts.flexOrder || true )
  // column padding, by default columns do not have padding
  // can conflict with $spacer if $spacer is set to 'padding'
  style.define( '$column-padding', padding || false )
  // whether or not to prefix the column classes (col-tab-2 vs tab-2)
  style.define( '$column-prefix', opts.columnPrefix || '' )
  // sets the max-width of the site
  style.define( '$site-width', opts.siteWidth || 80 )
  // set max width, in % or vw
  style.define( '$site-width-type', opts.widthType || '%' )
  // spacing between columns, margin or padding
  style.define( '$spacer', opts.spacer || 'margin' )
  // whether to put the margin/padding on the right or the left
  style.define( '$spacer-dir', opts.spacerDirection || 'right' )
  // unit type of spacing, defaults to percentage, can be any unit
  style.define( '$spacer-type', opts.spacerType || 'px' )
  // amount of spacing, number
  style.define( '$spacer-amount', opts.spacerAmount || 16 )
  // whether or not to include spacing on the bottom as well
  style.define( '$spacer-bottom', opts.spacerBottom || true )
  // generate mobile column classes (by default we do not, all columns are 100% on mobile)
  style.define( '$include-mobile', opts.includeMobile || true )
  // generate hide/show classes
  style.define( '$display-classes', opts.displayClasses || true )
  // generate remove/reveal classes
  style.define( '$display-overrides', opts.displayOverrides || true )
  // generate alpha classes
  style.define( '$column-alpha', opts.alpha || false )
  // generate omega classes
  style.define( '$column-omega', opts.omega || false )
  // generate push classes
  style.define( '$column-push', opts.push || false )
  // generate pull classes
  style.define( '$column-pull', opts.pull || false )

  // by default we don't generate gutters
  if ( typeof opts.gutters === 'object' ) {
    // if hard-coding gutters
    if ( opts.gutters.values ) {
      style.define( '$gutters', opts.gutters.values, true )
    }
    else {
      if ( opts.gutters.base ||
        opts.gutters.range ||
        opts.gutters.scale ) {
        // creates an array of gutters which we can use in stylus
        // generates based on a base value and a scale
        style.define( '$gutters', composeGutters( opts.gutters ), true )
      }
      else {
        // default gutters
        style.define( '$gutters', {
          0: 8,
          1: 16,
          2: 32,
          3: 64,
        }, true )
      }
    }

    // gutter variables, the margin and padding classes by breakpoint
    // gutter unit is the unit value for all gutters
    style.define( '$gutter-unit', opts.gutters.gutterUnit || 'px' )

    // below are values for auto generating gutters
    // the base gutter value, used as first gutter
    style.define( '$gutter-base', opts.gutters.base )
    // the number of gutter classes to create per direction per breakpoint
    style.define( '$gutter-range', opts.gutters.range )
    // the amount of increase for each gutter class
    style.define( '$gutter-scale', opts.gutters.scale )
  }
  else {
    // default gutters
    style.define( '$gutters', {
      0: 8,
      1: 16,
      2: 32,
      3: 64,
    }, true )
  }

  // directions hash, by default we generate
  // gutters for all 4 directions
  style.define( '$directions', {
    t: 'top',
    r: 'right',
    b: 'bottom',
    l: 'left',
  }, true )
}

module.exports = wiredStyle


lib/presets/toast.js
'use strict'

// @TODO would be interesting to see how feasible generating
// toast like '3-of-5' style grid systems dynamically would be
// https://daneden.github.io/Toast/
// depends on my laziness


/**
 * Module dependencies.
 */
var stylus = require( 'stylus' )
var _ = require( 'lodash' )
var nodes = stylus.nodes


// /**
//  * Return the plugin callback for stylus.
//  * @return {Function}
//  * @api public
//  */
// var toastGrid = function( style, opts ) {
//   // by default we support flexbox
//   var flexBox = true
//   if ( opts.flexSupport === false ) {
//     flexBox = false
//   }

//   // by default we don't include column padding
//   var padding = false
//   if ( opts.columnPadding ) {
//     padding =  new nodes.Literal( opts.columnPadding )
//   }

//   var defaultBreakpoints = {
//     small: {
//       max: 40,
//       isMobile: true,
//     },
//     medium: {
//       min: 40.063,
//     },
//     large: {
//       min: 64.063,
//     },
//     XLarge: {
//       min: 90.063,
//     },
//     XXLarge: {
//       min: 120.063,
//     }
//   }

//   // these are the breakpoints we use to define the grid
//   style.define( '$grid-breakpoints', opts.gridBreakpoints || defaultBreakpoints, true )
//   style.define( '$media-unit', opts.mediaUnit || 'rem' )

//   var mob = [{}]
//   if ( opts.gridBreakpoints ) {
//     mob = _.filter( opts.gridBreakpoints, function(b) {
//       return b.isMobile === true
//     } )
//   }
//   else {
//     mob = _.filter( defaultBreakpoints, function(b) {
//       return b.isMobile === true
//     } )
//   }

//   // we need to create a $mob-plus variable
//   var unit = opts.mediaUnit || 'rem'
//   var min = mob[0].max + 1
//   var mobPlus = '( min-width: ' +  min + unit + ' )'
//   style.define( '$mob-plus', mobPlus )

//   // - GRID SETTINGS
//   // how many columns to generate
//   style.define( '$columns', opts.columns || 12 )
//   // base column class
//   style.define( '$column-base', opts.columnBaseClass || 'column' )

//   // basic flex box support?
//   style.define( '$flexbox-support', flexBox )
//   // generate flexbox re-ordering classes
//   style.define( '$flexbox-order', opts.flexOrder || true )
//   // column padding, by default columns do not have padding
//   // can conflict with $spacer if $spacer is set to 'padding'
//   style.define( '$column-padding', padding || false )
//   // whether or not to prefix the column classes (col-tab-2 vs tab-2)
//   style.define( '$column-prefix', opts.columnPrefix || '' )
//   // sets the max-width of the site
//   style.define( '$site-width', opts.siteWidth || 72.25 )
//   // set max width, in % or vw or em or rem
//   style.define( '$site-width-type', opts.widthType || 'rem' )
//   // spacing between columns, margin or padding
//   style.define( '$spacer', opts.spacer || 'padding' )
//   // whether to put the margin/padding on the right or the left or both
//   style.define( '$spacer-dir', opts.spacerDirection || 'both' )
//   // unit type of spacing, defaults to percentage, can be any unit
//   style.define( '$spacer-type', opts.spacerType || 'rem' )
//   // amount of spacing, number
//   style.define( '$spacer-amount', opts.spacerAmount || 0.9375 )
//   // whether or not to include spacing on the bottom as well
//   style.define( '$spacer-bottom', opts.spacerBottom || false )
//   // generate mobile column classes (by default we do not, all columns are 100% on mobile)
//   style.define( '$include-mobile', opts.includeMobile || true )
//   // generate hide/show classes
//   style.define( '$display-classes', opts.displayClasses || true )
//   // generate remove/reveal classes
//   style.define( '$display-overrides', opts.displayOverrides || false )
//   // generate alpha classes
//   style.define( '$column-alpha', opts.alpha || true )
//   // generate omega classes
//   style.define( '$column-omega', opts.omega || true )
//   // generate push classes
//   style.define( '$column-push', opts.push || true )
//   // generate pull classes
//   style.define( '$column-pull', opts.pull || true )

//   // directions hash, by default we generate
//   // gutters for all 4 directions
//   style.define( '$directions', {
//     t: 'top',
//     r: 'right',
//     b: 'bottom',
//     l: 'left',
//   }, true )
// }


// module.exports = foundationGrid


lib/composure-grid.js
'use strict'

var stylus = require( 'stylus' )

// default grid, 12 columns, 2 breakpoints 3 gutters hide/show, flexbox/order
var simpleGrid = require( './presets/simple' )

// optional presets
// WIRED style
// 18 columns 3 breakpoints 5 gutters hide/show remove/reveal, flexbox/order, col- prefix
var wiredStyle = require( './presets/WIRED' )

// bootstrap style
// 12 columns 5 breakpoints no gutters hide/show, flexbox/order, col- prefix
var bootStyle = require( './presets/bootstrap' )

// bootstrap style
// 12 columns 5 breakpoints no gutters hide/show, flexbox/order, col- prefix
var foundationStyle = require( './presets/foundation' )


/**
 * Return the plugin callback for stylus.
 * @return {Function}
 * @api public
 */
var composureGrid = function( opts ) {
  return function( style ) {
    style.include( __dirname )

    // if no opts passed in just create an empty obj
    if ( typeof opts !== 'object' ) {
      opts = {}
    }

    if ( opts.preset === 'WIRED'.toLowerCase() ) {
      return wiredStyle( style, opts )
    }

    if ( opts.preset === 'bootstrap' ) {
      return bootStyle( style, opts )
    }

    if ( opts.preset === 'foundation' ) {
      return foundationStyle( style, opts )
    }

    return simpleGrid( style, opts )
  }
}


exports = module.exports = composureGrid


lib/presets/simple.js
'use strict'


/**
 * Module dependencies.
 */
var stylus = require( 'stylus' )
var _ = require( 'lodash' )
var nodes = stylus.nodes

// gutter maker helper function
var composeGutters = require( '../utils/gutters' )


/**
 * Return the plugin callback for stylus.
 * @return {Function}
 * @api public
 */
var simpleGrid = function( style, opts ) {
  // by default we support flexbox
  var flexBox = true
  if ( opts.flexSupport === false ) {
    flexBox = false
  }

  // by default we don't include column padding
  var padding = false
  if ( opts.columnPadding ) {
    padding =  new nodes.Literal( opts.columnPadding )
  }

  var defaultBreakpoints = {
    mob: {
      max: 767,
      isMobile: true,
    },
    tab: {
      min: 768,
    },
    desk: {
      min: 1200,
    },
  }

  // these are the breakpoints we use to define the grid
  style.define( '$grid-breakpoints', opts.gridBreakpoints || defaultBreakpoints, true )
  style.define( '$media-unit', opts.mediaUnit || 'px' )

  var mob = [{}]
  if ( opts.gridBreakpoints ) {
    mob = _.filter( opts.gridBreakpoints, function(b) {
      return b.isMobile === true
    } )
  }
  else {
    mob = _.filter( defaultBreakpoints, function(b) {
      return b.isMobile === true
    } )
  }

  // we need to create a $mob-plus variable
  var unit = opts.mediaUnit || 'px'
  var min = mob[0].max + 1
  var mobPlus = '( min-width: ' +  min + unit + ' )'
  style.define( '$mob-plus', mobPlus )

  // - GRID SETTINGS
  // how many columns to generate
  style.define( '$columns', opts.columns || 12 )
  // base column class
  style.define( '$column-base', opts.columnBaseClass || 'col' )

  // basic flex box support?
  style.define( '$flexbox-support', flexBox )
  // generate flexbox re-ordering classes
  style.define( '$flexbox-order', opts.flexOrder || true )
  // column padding, by default columns do not have padding
  // can conflict with $spacer if $spacer is set to 'padding'
  style.define( '$column-padding', padding || false )
  // whether or not to prefix the column classes (col-tab-2 vs tab-2)
  style.define( '$column-prefix', opts.columnPrefix || '' )
  // sets the max-width of the site
  style.define( '$site-width', opts.siteWidth || 80 )
  // set max width, in % or vw
  style.define( '$site-width-type', opts.widthType || '%' )
  // spacing between columns, margin or padding
  style.define( '$spacer', opts.spacer || 'margin' )
  // whether to put the margin/padding on the right or the left
  style.define( '$spacer-dir', opts.spacerDirection || 'right' )
  // unit type of spacing, defaults to percentage, can be any unit
  style.define( '$spacer-type', opts.spacerType || 'px' )
  // amount of spacing, number
  style.define( '$spacer-amount', opts.spacerAmount || 20 )
  // whether or not to include spacing on the bottom as well
  style.define( '$spacer-bottom', opts.spacerBottom || true )
  // generate mobile column classes (by default we do not, all columns are 100% on mobile)
  style.define( '$include-mobile', opts.includeMobile || false )
  // generate hide/show classes
  style.define( '$display-classes', opts.displayClasses || true )
  // generate remove/reveal classes
  style.define( '$display-overrides', opts.displayOverrides || false )
  // generate alpha classes
  style.define( '$column-alpha', opts.alpha || false )
  // generate omega classes
  style.define( '$column-omega', opts.omega || true )
  // generate push classes
  style.define( '$column-push', opts.push || false )
  // generate pull classes
  style.define( '$column-pull', opts.pull || false )

  // by default we don't generate gutters
  if ( opts.gutters ) {
    if ( typeof opts.gutters === 'object' ) {
      // if hard-coding gutters
      if ( opts.gutters.values ) {
        style.define( '$gutters', opts.gutters.values, true )
      }
      else {
        if ( opts.gutters.base ||
          opts.gutters.range ||
          opts.gutters.scale ) {
          // creates an array of gutters which we can use in stylus
          // generates based on a base value and a scale
          style.define( '$gutters', composeGutters( opts.gutters ), true )
        }
      }

      // gutter variables, the margin and padding classes by breakpoint
      // gutter unit is the unit value for all gutters
      style.define( '$gutter-unit', opts.gutters.gutterUnit || 'px' )

      // below are values for auto generating gutters
      // the base gutter value, used as first gutter
      style.define( '$gutter-base', opts.gutters.base )
      // the number of gutter classes to create per direction per breakpoint
      style.define( '$gutter-range', opts.gutters.range )
      // the amount of increase for each gutter class
      style.define( '$gutter-scale', opts.gutters.scale )
    }
    else {
      // default gutters if just set to true
      style.define( '$gutters', {
        0: 10,
        1: 20,
        2: 40,
      }, true )
    }
  }

  // directions hash, by default we generate
  // gutters for all 4 directions
  style.define( '$directions', {
    t: 'top',
    r: 'right',
    b: 'bottom',
    l: 'left',
  }, true )
}


module.exports = simpleGrid


lib/presets/bootstrap.js
'use strict'


/**
 * Module dependencies.
 */
var stylus = require( 'stylus' )
var _ = require( 'lodash' )
var nodes = stylus.nodes


/**
 * Return the plugin callback for stylus.
 * @return {Function}
 * @api public
 */
var bootstrapGrid = function( style, opts ) {
  // by default we support flexbox
  var flexBox = true
  if ( opts.flexSupport === false ) {
    flexBox = false
  }

  // by default we don't include column padding
  var padding = false
  if ( opts.columnPadding ) {
    padding =  new nodes.Literal( opts.columnPadding )
  }

  var defaultBreakpoints = {
    xs: {
      max: 33.99,
      isMobile: true,
    },
    sm: {
      min: 34,
    },
    md: {
      min: 48,
    },
    lg: {
      min: 62,
    },
    xl: {
      min: 75,
    }
  }

  // these are the breakpoints we use to define the grid
  style.define( '$grid-breakpoints', opts.gridBreakpoints || defaultBreakpoints, true )
  style.define( '$media-unit', opts.mediaUnit || 'rem' )

  var mob = [{}]
  if ( opts.gridBreakpoints ) {
    mob = _.filter( opts.gridBreakpoints, function(b) {
      return b.isMobile === true
    } )
  }
  else {
    mob = _.filter( defaultBreakpoints, function(b) {
      return b.isMobile === true
    } )
  }

  // we need to create a $mob-plus variable
  var unit = opts.mediaUnit || 'rem'
  var min = mob[0].max + 1
  var mobPlus = '( min-width: ' +  min + unit + ' )'
  style.define( '$mob-plus', mobPlus )

  // - GRID SETTINGS
  // how many columns to generate
  style.define( '$columns', opts.columns || 12 )
  // base column class
  style.define( '$column-base', opts.columnBaseClass || 'col' )

  // basic flex box support?
  style.define( '$flexbox-support', flexBox )
  // generate flexbox re-ordering classes
  style.define( '$flexbox-order', opts.flexOrder || true )
  // column padding, by default columns do not have padding
  // can conflict with $spacer if $spacer is set to 'padding'
  style.define( '$column-padding', padding || false )
  // whether or not to prefix the column classes (col-tab-2 vs tab-2)
  style.define( '$column-prefix', opts.columnPrefix || 'col-' )
  // sets the max-width of the site
  style.define( '$site-width', opts.siteWidth || 72.25 )
  // set max width, in % or vw or em or rem
  style.define( '$site-width-type', opts.widthType || 'rem' )
  // spacing between columns, margin or padding
  style.define( '$spacer', opts.spacer || 'padding' )
  // whether to put the margin/padding on the right or the left or both
  style.define( '$spacer-dir', opts.spacerDirection || 'both' )
  // unit type of spacing, defaults to percentage, can be any unit
  style.define( '$spacer-type', opts.spacerType || 'px' )
  // amount of spacing, number
  style.define( '$spacer-amount', opts.spacerAmount || 15 )
  // whether or not to include spacing on the bottom as well
  style.define( '$spacer-bottom', opts.spacerBottom || false )
  // generate mobile column classes (by default we do not, all columns are 100% on mobile)
  style.define( '$include-mobile', opts.includeMobile || true )
  // generate hide/show classes
  style.define( '$display-classes', opts.displayClasses || true )
  // generate remove/reveal classes
  style.define( '$display-overrides', opts.displayOverrides || false )
  // generate alpha classes
  style.define( '$column-alpha', opts.alpha || true )
  // generate omega classes
  style.define( '$column-omega', opts.omega || true )
  // generate push classes
  style.define( '$column-push', opts.push || true )
  // generate pull classes
  style.define( '$column-pull', opts.pull || true )

  // directions hash, by default we generate
  // gutters for all 4 directions
  style.define( '$directions', {
    t: 'top',
    r: 'right',
    b: 'bottom',
    l: 'left',
  }, true )
}


module.exports = bootstrapGrid


lib/presets/foundation.js
'use strict'


/**
 * Module dependencies.
 */
var stylus = require( 'stylus' )
var _ = require( 'lodash' )
var nodes = stylus.nodes


/**
 * Return the plugin callback for stylus.
 * @return {Function}
 * @api public
 */
var foundationGrid = function( style, opts ) {
  // by default we support flexbox
  var flexBox = true
  if ( opts.flexSupport === false ) {
    flexBox = false
  }

  // by default we don't include column padding
  var padding = false
  if ( opts.columnPadding ) {
    padding =  new nodes.Literal( opts.columnPadding )
  }

  var defaultBreakpoints = {
    small: {
      max: 40,
      isMobile: true,
    },
    medium: {
      min: 40.063,
    },
    large: {
      min: 64.063,
    },
    XLarge: {
      min: 90.063,
    },
    XXLarge: {
      min: 120.063,
    }
  }

  // these are the breakpoints we use to define the grid
  style.define( '$grid-breakpoints', opts.gridBreakpoints || defaultBreakpoints, true )
  style.define( '$media-unit', opts.mediaUnit || 'rem' )

  var mob = [{}]
  if ( opts.gridBreakpoints ) {
    mob = _.filter( opts.gridBreakpoints, function(b) {
      return b.isMobile === true
    } )
  }
  else {
    mob = _.filter( defaultBreakpoints, function(b) {
      return b.isMobile === true
    } )
  }

  // we need to create a $mob-plus variable
  var unit = opts.mediaUnit || 'rem'
  var min = mob[0].max + 1
  var mobPlus = '( min-width: ' +  min + unit + ' )'
  style.define( '$mob-plus', mobPlus )

  // - GRID SETTINGS
  // how many columns to generate
  style.define( '$columns', opts.columns || 12 )
  // base column class
  style.define( '$column-base', opts.columnBaseClass || 'column' )

  // basic flex box support?
  style.define( '$flexbox-support', flexBox )
  // generate flexbox re-ordering classes
  style.define( '$flexbox-order', opts.flexOrder || true )
  // column padding, by default columns do not have padding
  // can conflict with $spacer if $spacer is set to 'padding'
  style.define( '$column-padding', padding || false )
  // whether or not to prefix the column classes (col-tab-2 vs tab-2)
  style.define( '$column-prefix', opts.columnPrefix || '' )
  // sets the max-width of the site
  style.define( '$site-width', opts.siteWidth || 72.25 )
  // set max width, in % or vw or em or rem
  style.define( '$site-width-type', opts.widthType || 'rem' )
  // spacing between columns, margin or padding
  style.define( '$spacer', opts.spacer || 'padding' )
  // whether to put the margin/padding on the right or the left or both
  style.define( '$spacer-dir', opts.spacerDirection || 'both' )
  // unit type of spacing, defaults to percentage, can be any unit
  style.define( '$spacer-type', opts.spacerType || 'rem' )
  // amount of spacing, number
  style.define( '$spacer-amount', opts.spacerAmount || 0.9375 )
  // whether or not to include spacing on the bottom as well
  style.define( '$spacer-bottom', opts.spacerBottom || false )
  // generate mobile column classes (by default we do not, all columns are 100% on mobile)
  style.define( '$include-mobile', opts.includeMobile || true )
  // generate hide/show classes
  style.define( '$display-classes', opts.displayClasses || true )
  // generate remove/reveal classes
  style.define( '$display-overrides', opts.displayOverrides || false )
  // generate alpha classes
  style.define( '$column-alpha', opts.alpha || true )
  // generate omega classes
  style.define( '$column-omega', opts.omega || true )
  // generate push classes
  style.define( '$column-push', opts.push || true )
  // generate pull classes
  style.define( '$column-pull', opts.pull || true )

  // directions hash, by default we generate
  // gutters for all 4 directions
  style.define( '$directions', {
    t: 'top',
    r: 'right',
    b: 'bottom',
    l: 'left',
  }, true )
}


module.exports = foundationGrid


gulpfile.js
'use strict'

var gulp = require( 'gulp' )
var stylus = require( 'gulp-stylus' )
var mq = require( 'css-mqpacker' )
var postcss = require( 'gulp-postcss' )
var composure = require( './lib/composure-grid' )

/**
 * compiles, autoprefixes, minifies, the stylus
 */
gulp.task( 'stylus', function() {
  return gulp.src( './lib/composure-grid.styl' )
    .pipe( stylus( {
      compress: true,
      use: composure( {
        alpha: false,
        columns: 12,
        columnBaseClass: 'col',
        columnPadding: false,
        columnPrefix: '',
        directions: {
          t: 'top',
          r: 'right',
          b: 'bottom',
          l: 'left',
        },
        displayClasses: true,
        displayOverrides: false,
        flexSupport: true,
        flexOrder: true,
        gridBreakpoints: {
          m: {
            max: 767,
            isMobile: true,
          },
          t: {
            min: 768,
          },
          d: {
            min: 1200,
          },
        },
        gutters: {
          base: 10,
          range: 3,
          scale: 2,
          values: {
            0: 10,
            1: 20,
            2: 40,
          },
        },
        includeMobile: false,
        mediaUnit: 'px',
        omega: true,
        preset: 'simple',
        pull: false,
        push: false,
        siteWidth: 80,
        spacer: 'margin',
        spacerAmount: 1,
        spacerBottom: true,
        spacerDirection: 'right',
        spacerType: '%',
        widthType: '%',
      } ),
    } ) )
    .pipe( postcss( [mq] ) )
    .pipe( gulp.dest( './css' ) )
} )

/**
 * Watch for file changes and rebuild. For client side changes we can
 * automatically reload the files without having to reload the server.
 * Server side changes require a SIGHUP to restart the server.
 */
gulp.task('watch', function() {
  return gulp.watch(['./styl/**/*.styl'], ['stylus'])
})


gulp.task( 'default', ['stylus', 'watch'] )
